keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlQueue.kt

```diff
@@ -0,0 +1,743 @@
+  override val publisher: EventPublisher,
+  private val sqlRetryProperties: SqlRetryProperties
+) : MonitorableQueue {
+
+  companion object {
+    @Suppress("UnstableApiUsage")
+    /**
+     * [lockId] is a hash of the hostname, used to claim messages on [queueTable] without
+     * performing a locking read.
+     */
+    private val lockId = Hashing
+      .murmur3_128()
+      .hashString(InetAddress.getLocalHost().hostName, StandardCharsets.UTF_8)
+      .toString()
+
+    private val hashObjectMapper = ObjectMapper().copy().apply {
+      enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS)
+    }
+
+    private val lockedAtRegex = """^\w+:(\d+)$""".toRegex()
+    private val nameSanitization = """[^A-Za-z0-9_]""".toRegex()
+
+    private val log = LoggerFactory.getLogger(SqlQueue::class.java)
+  }
+
+  private val sanitizedName = queueName.replace(nameSanitization, "_")
+  private val queueBase = "keiko_v${schemaVersion}_queue"
+  private val unackedBase = "keiko_v${schemaVersion}_unacked"
+  private val messagesBase = "keiko_v${schemaVersion}_messages"
+
+  private val queueTableName = "${queueBase}_$sanitizedName"
+  private val unackedTableName = "${unackedBase}_$sanitizedName"
+  private val messagesTableName = "${messagesBase}_$sanitizedName"
+
+  private val queueTable = table(queueTableName)
+  private val unackedTable = table(unackedTableName)
+  private val messagesTable = table(messagesTableName)
+
+  private val bodyField = field("body")
+  private val deliveryField = field("delivery")
+  private val expiryField = field("expiry")
+  private val fingerprintField = field("fingerprint")
+  private val idField = field("id")
+  private val lockedField = field("locked")
+
+  private val lockTtlDuration = Duration.ofSeconds(lockTtlSeconds.toLong())
+
+  init {
+    log.info("Configured $javaClass queue: $queueName")
+    initTables()
+  }
+
+  override fun readState(): QueueState {
+    /**
+     * Reading counts across all tables in a single query for consistency
+     */
+    val rs = jooq.select()
+      .select(
+        select(count())
+          .from(queueTable)
+          .asField<Int>("depth"),
+        select(count())
+          .from(queueTable)
+          .where(deliveryField.le(clock.instant().toEpochMilli()))
+          .asField<Int>("ready"),
+        select(count())
+          .from(unackedTable)
+          .asField<Int>("unacked"),
+        select(count())
+          .from(messagesTable)
+          .asField<Int>("messages")
+      )
+      .fetch()
+      .intoResultSet()
+
+    rs.next()
+
+    val depth = rs.getInt("depth")
+    val ready = rs.getInt("ready")
+    val unacked = rs.getInt("unacked")
+    val messages = rs.getInt("messages")
+
+    return QueueState(
+      depth = depth,
+      ready = ready,
+      unacked = unacked,
+      orphaned = messages - (depth + unacked)
+    )
+  }
+
+  override fun containsMessage(predicate: (Message) -> Boolean): Boolean {
+    val limit = 100
```

Comment:
This strikes me as a pretty shallow search. Our steady-state queue depth is ~1.2k. Why 100? Would this be worthwhile exposing to configuration?

Reply:
Oh. I see now, this is a batch size, but we go through everything. Maybe change this to `batchSize` instead?